New RPGLoot Affixes + Component Additions

Whoneedspacee opened the pull request:
## About The Pull Request

Adds fired_from to projectiles to store the thing the projectile was fired from, firer = human that fired, fired_from = the gun that shot the projectile. Knockback now handles projectiles knocking back and also anchored structures being thrown.

Adds shrapnel component, a component that shoots projectiles in a radius around a projectile when it hits something. Adds igniter component, sets mobs the parent hits on fire. Adds summoning component, summons mobs when the parent hits something. Adds lifesteal component that attempts to heal the attacker for an amount when they hit something living.

Also adds some ishelpers and signals

## Why It's Good For The Game

RPGLoot is pretty lacking in actual components that do anything and for the most part are completely cosmetic. This doesn't totally address the imbalance of cosmetic affixes that do absolutely nothing, but at least adds some more variety other than 2 things that can occur.

## Changelog
:cl:
add: New RPGLoot modifiers: Vampirism which heals you when you attack, Pyromantic which sets things you hit on fire. Shrapnel which causes projectiles fired from a gun to fire projectiles in a radius when they hit something. Finally, Summoning which summons mobs that sometimes aid you in combat.
/:cl:

Whoneedspacee committed:
knockback component can now be reversed, has projectile and gun handling, and hostile simplemob handling

adds signals for hostile mobs attacking, altering projectiles before firing, and for when projectiles successfully hit their target

moves knockback handling to a general proc

adds ishelpers for guns and projectiles

Whoneedspacee committed:
no more weird projectile handling it can just not apply the effect if the component somehow goes away

lifesteal actually works now instead of being a blank file, applies a flat healing effect when you hit something

Whoneedspacee committed:
fixes up comsig stuff

adds new components to the fantasy prefix and suffix

knockback now handles throwing anchored objects

lifesteal now properly heals the target with projectiel weapons

adds summoning component to handle mob summoning with item attacking and such

adds fired_from variable to handle what a projectile was fired_from, firer would be the mob that fired and fired_from would be the gun, in the case of an autoturret, fired_from and firer would be the same

adds shrapnel component, fires projectiles around a fired projectile when it hits

adds igniter component to set attacked mobs on fire

optimumtact commented:
![](https://file.house/ipLF.jpg)

Militaires commented:
good pr

Razharas commented:
yeah but what about adding the stats and resistances to weapons and armour?
how am i supposed to fight high tier boss content a if only options are for white/blue clearspeed?

4dplanner commented:
If there are going to be more of these fantasy components, maybe they should get their own folder?

Whoneedspacee commented:
they aren’t specifically for fantasy they are components.

the fantasy component already has its own folder which applies these components in the specific way they should be for rpgloot.

4dplanner commented:
lol I actually just misread the path on the fantasy components, ignore me

ninjanomnom commented on code/datums/components/fantasy/suffixes.dm:
```diff
@@ -78,14 +77,76 @@
+	var/static/list/possible_projectiletypes
+	if(!possible_projectiletypes)
+		// The base list of allowed types
+		possible_projectiletypes = typecacheof(list(
```
This should probably be hand picked and not just every kind, they have a *lot* of variety and many will break things.

This also lets you weight them if you want

ninjanomnom commented on code/datums/components/knockback.dm:
```diff
@@ -1,22 +1,45 @@
 /datum/component/knockback
 	var/throw_distance
+	var/reverse
```
Just have a negative throw distance?

Whoneedspacee replied:
oh duh haha i got stuck up in the fact that getting negative target turfs wouldnt work

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into newfantasycomponents

Whoneedspacee committed:
no more shrapnel on items that can't use it

summoning items now summon at least one mob maximum

adds specific weighted projectile types for shrapnel to prevent broken options being picked

removes the reverse var from knockback component and instead just handles negative thrown turf
